File: platforms/android/samples/CheckoutKitAndroidDemo/app/src/main/java/com/shopify/checkout_kit_android_demo/e2e/E2EController.kt
package com.shopify.checkout_kit_android_demo.e2e

interface E2ECommandTarget {
    suspend fun selectBuyerIdentityMode(mode: E2EBuyerIdentityMode)

    suspend fun resetCart()

    suspend fun variantId(atProductIndex: Int): String

    suspend fun addCartLine(variantId: String, quantity: Int)

    suspend fun showCart()

    suspend fun report(failure: String)
}

class E2EController(private val target: E2ECommandTarget) {

    suspend fun handle(url: String): Boolean {
        val link = try {
            E2EControlLink.parse(url)
        } catch (error: IllegalArgumentException) {
            target.report(message(error))
            return true
        }

        if (link == null) {
            return false
        }

        perform(link)

        return true
    }

    private suspend fun perform(link: E2EControlLink) {
        try {
            when (link) {
                is E2EControlLink.Reset -> target.resetCart()
                is E2EControlLink.Cart -> seedCart(link)
                is E2EControlLink.SignIn -> throw UnsupportedOperationException("signIn is not implemented yet")
            }
        } catch (error: Exception) {
            target.report(message(error))
        }
    }

    private suspend fun seedCart(command: E2EControlLink.Cart) {
        command.buyerIdentityMode?.let { target.selectBuyerIdentityMode(it) }

        target.resetCart()

        val variantId = command.variantId ?: target.variantId(command.productIndex ?: 0)

        target.addCartLine(variantId, command.quantity)
        target.showCart()
    }

    private fun message(error: Throwable) = error.message ?: error.toString()
}

File: platforms/swift/Samples/CheckoutKitSwiftDemo/CheckoutKitSwiftDemo/Sources/E2E/E2EController.swift
import Foundation

@MainActor
protocol E2ECommandTarget {
    func selectBuyerIdentityMode(_ mode: BuyerIdentityMode) async
    func resetCart() async
    func variantId(atProductIndex index: Int) async throws -> String
    func addCartLine(variantId: String, quantity: Int) async throws
    func showCart() async
    func presentSignIn() async
    func report(failure message: String) async
}

enum E2EControllerError: LocalizedError, Equatable {
    case productIndexOutOfRange(Int)

    var errorDescription: String? {
        switch self {
        case let .productIndexOutOfRange(index):
            return "No product at index \(index)"
        }
    }
}

@MainActor
final class E2EController {
    static let shared = E2EController(target: E2ESampleAppTarget())

    private let target: E2ECommandTarget

    /// Chains every perform(_:) onto the previous one, because @MainActor alone does not
    /// stop a second handle(url:) call from interleaving with the first across await points.
    private var tail: Task<Void, Never>?

    init(target: E2ECommandTarget) {
        self.target = target
    }

    @discardableResult
    func handle(url: String) async -> Bool {
        let link: E2EControlLink?

        do {
            link = try E2EControlLink.parse(url)
        } catch {
            await target.report(failure: E2EController.message(for: error))
            return true
        }

        guard let link else {
            return false
        }

        await enqueue(link)

        return true
    }

    private func enqueue(_ link: E2EControlLink) async {
        let previous = tail
        let task = Task {
            await previous?.value
            await self.perform(link)
        }
        tail = task
        await task.value
    }

    private func perform(_ link: E2EControlLink) async {
        do {
            switch link {
            case .reset:
                await target.resetCart()
            case let .cart(command):
                try await seedCart(command)
            case .signIn:
                await target.presentSignIn()
            }
        } catch {
            await target.report(failure: E2EController.message(for: error))
        }
    }

    private func seedCart(_ command: E2EControlLink.CartCommand) async throws {
        if let buyerIdentityMode = command.buyerIdentityMode {
            await target.selectBuyerIdentityMode(buyerIdentityMode)
        }

        await target.resetCart()

        let variantId = try await variantId(for: command)

        try await target.addCartLine(variantId: variantId, quantity: command.quantity)
        await target.showCart()
    }

    private func variantId(for command: E2EControlLink.CartCommand) async throws -> String {
        if let variantId = command.variantId {
            return variantId
        }

        return try await target.variantId(atProductIndex: command.productIndex ?? 0)
    }

    private static func message(for error: Error) -> String {
        (error as? LocalizedError)?.errorDescription ?? error.localizedDescription
    }
}
